### DISCUSSION TITLE
updateConfig in integration hook should return updated config

### DISCUSSION THREAD

**Discussion by nerdoza** [withastro/roadmap#762] [Proposal]
### Body

# Summary

In the `astro:config:setup` there is a utility to update the configuration (`updateConfig`) using a `mergeConfig` function, but the resulting config is not returned as part of the function so code within the hook needs to track changes independently.

# Background & Motivation

The `config` should be the source of truth and not require integration implementors to guess at the functionality of the `mergeConfig` internal.

# Goals

The ideal resolution is to be able to update the in scope `config` to the match the internal `updatedConfig` value.
```ts
"astro:config:setup": ({ config, updateConfig }) => {
  let updatedConfig = config

  if (shouldForceStaticOutput) {
    updatedConfig = updateConfig({ output: "static" })
  }
  
  if (updatedConfig.output === "static") {
    // Disable sourcemaps for static site
    updatedConfig = updateConfig({ vite: { build: { sourcemap: true } } })
  }
}
```

# Example

Add a return to the `updateConfig` method.
 https://github.com/withastro/astro/blob/35739d01e9cc4fa31a8b85201feecf29c747eca9/packages/astro/src/integrations/index.ts#L119-L121

```ts
updateConfig: (newConfig) => {
  updatedConfig = mergeConfig(updatedConfig, newConfig) as AstroConfig;
  return updatedConfig;
},
```

**Comment by bluwy** [comment]
Based on the examples shown, I don't quite understand why this feature is needed yet. You're controlling when and how `updateConfig` is called, so you would have context if a value is already set a certain way, and don't need to re-read the new updated config again? For example, you'd know `output === 'static'` based on `shouldForceStaticOutput`.

**Reply by nerdoza** [reply]
The `output` might have been set to `static` by the user in the configuration and `shouldForceStaticOutput` could be `false`.  To solve this same issue in the current implementation requires tracking the changes manually.

```ts
// Current Implementation Strategy
{
  "astro:config:setup": ({ config, updateConfig }) => {
    let updatedConfig = config
  
    if (shouldForceStaticOutput) {
      // Override output mode if adapter is configured for SSG only
      updateConfig({ output: "static" })
      updatedConfig.output = "static"
    }
    
    if (updatedConfig.output === "static") {
      // Disable sourcemaps for static sites
      updateConfig({ vite: { build: { sourcemap: true } } })
      updatedConfig.vite.build.souremap = true;
    }

    if (updatedConfig.vite.build.souremap) {
      // Do something else based on if souremaps need to be generated
    }
  }
}
```
This example is a bit trivial, but hopefully it helps to explain the issue a bit more. I suspect the intention of the `updateConfig` call was for it to be called only once from the hook, but in practice we've found it useful to be able to apply `updateConfig` in stages, but the state logic gets ugly fast without being able to rely on the actual `config` for source of truth.

A PR to implement this feature is available here: https://github.com/withastro/astro/pull/9013#issuecomment-1797918274

**Comment by nerdoza** [comment]
PR with implementation merged (https://github.com/withastro/astro/pull/9013#issuecomment-1797918274) as part of `v4` release.
